sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala

```diff
@@ -414,16 +417,16 @@ class ParquetFileFormat
         } else {
           new ParquetRecordReader[UnsafeRow](new ParquetReadSupport(convertTz))
         }
+        val recordReaderIterator = new RecordReaderIterator(reader)
+        // Register a task completion lister before `initalization`.
+        taskContext.foreach(_.addTaskCompletionListener(_ => recordReaderIterator.close()))
         reader.initialize(split, hadoopAttemptContext)
-        reader
+        recordReaderIterator
       }
 
-      val iter = new RecordReaderIterator(parquetReader)
-      taskContext.foreach(_.addTaskCompletionListener(_ => iter.close()))
 
       // UnsafeRowParquetRecordReader appends the columns internally to avoid another copy.
-      if (parquetReader.isInstanceOf[VectorizedParquetRecordReader] &&
-          enableVectorizedReader) {
+      if (enableVectorizedReader) {
```

Comment:
Would it be possible to merge this if-statement into the above if-statement?

Reply:
Yep. It looks possible. I'll update together after getting more reviews. Thanks, @kiszk .

Reply:
yea it seems more reasonable to merge this if-else now.